src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+            first = newNode;
+        } else {
+            newlast.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            final Node<T> pred = node(index).prev;
+            final Node<T> newNode = new Node<>(pred, value, node(index));
+            node(index).prev = newNode;
+            if (pred == null) {
+                first = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size();i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> rewrite = node(index);
+        T oldValue = rewrite.item;
+        rewrite.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index);
+        return removeByObject(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> i = first; i != null; i = i.next) {
+                if (i.item == null) {
+                    removeByObject(i);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> i = first; i != null; i = i.next) {
+                if (object.equals(i.item)) {
+                    removeByObject(i);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> nodeByIndex = first;
+            for (int i = 0; i < index; i++) {
+                nodeByIndex = nodeByIndex.next;
+            }
+            return nodeByIndex;
+        } else {
+            Node<T> nodeByIndex = last;
+            for (int i = size - 1; i > index; i--) {
+                nodeByIndex = nodeByIndex.prev;
+            }
+            return nodeByIndex;
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void checkPositionIndexForAdd(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
```

Comment:
Why we need almost the same method? And if you add this method to node you can avoid redurant code